astropy/wcs/utils.py

```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
+                        projection='TAN', sip_degree=4):
+
+    """ 
+    Given two matching sets of coordinates on detector and sky,
+    compute the WCS. 
+
+    Fits a WCS object to matched set of input detector and sky coordinates.
+    Optionally, a SIP can be fit to account for geometric
+    distortion. Returns an `~astropy.wcs.WCS` object with the best fit
+    parameters for mapping between input pixel and sky coordinates.
+
+    The projection type (default 'TAN') can passed in as a string, one of 
+    the valid three-letter projection codes - or as a WCS object with
+    projection keywords already set. Note that if an input WCS has any 
+    non-polynomial distortion, this will be applied and reflected in the 
+    fit terms and coefficients. Passing in a WCS object in this way essentially 
+    allows it to be refit based on the matched input coordinates and projection 
+    point, but take care when using this option as non-projection related 
+    keywords in the input might cause unexpected behavior. 
+
+    Notes
+    ------
+    - The fiducial point for the spherical projection can be set to 'center'
+      to use the mean position of input sky coordinates, or as an
+      `~astropy.coordinates.SkyCoord` object.
+    - All output WCS will always be in degrees.
+    - If the coordinate frame differs between `~astropy.coordinates.SkyCoord`
+      objects passed in for `world_coords` and `proj_point`, the frame for
+      `world_coords`  will override as the frame for the output WCS.
+
+    Parameters
+    ----------
+    xy : tuple of two `numpy.ndarray`
+        x & y pixel coordinates.
+    world_coords : `~astropy.coordinates.SkyCoord`
+        Skycoord object with world coordinates.
+    proj_point : 'center' or ~astropy.coordinates.SkyCoord`
+        Defaults to 'center', in which the geometric center of input world
+        coordinates will be used as the projection point. To specify an exact
+        point for the projection, a Skycoord object with a coordinate pair can
+        be passed in. For consistency, the units and frame of these coordinates
+        will be transformed to match 'world_coords' if they don't.
+    projection : str or `~astropy.wcs.WCS`
+        Three letter projection code, of any of standard projections defined 
+        in the FITS WCS standard. Optionally, a WCS object with projection 
+        keywords set may be passed in. 
+    sip_degree : int
+        If set to a non-zero integer value, will fit SIP of degree `sip_degree` 
+        to model geometric distortion. 
+
+    Returns
+    -------
+    wcs : `~astropy.wcs.WCS`
+        The best-fit WCS to the points given.
+    """
+
+    from astropy.coordinates import SkyCoord # here to avoid circular import
+    import astropy.units as u
+    from scipy.optimize import least_squares
+    from .wcs import Sip
+    import copy
+
+    xp, yp = xy
+    lon, lat = world_coords.data.lon.deg, world_coords.data.lat.deg
```

Comment:
I don't think you can assume that these will always use a spherical representation - maybe you put this in a ``try...except`` and if an ``AttributeError`` is raised you can access:

```
unit_sph =  world_coords.unit_spherical
lon, lat = unit_sph.lon.deg, unit_sph.lat.deg
```